crates/sui-bridge-indexer/src/eth_bridge_indexer.rs

```diff
@@ -0,0 +1,203 @@
+
+use crate::indexer_builder::{
+    DataFilter, DataMapper, Datasource, IndexerProgressStore, Persistent,
+};
+use crate::sui_bridge_indexer::PgBridgePersistent;
+use crate::{
+    BridgeDataSource, ProcessedTxnData, TokenTransfer, TokenTransferData, TokenTransferStatus,
+};
+
+type EthData = (EthLog, Block<H256>, Transaction);
+
+pub struct EthFinalizedDatasource {
+    bridge_address: EthAddress,
+    eth_rpc_url: String,
+    bridge_metrics: Arc<BridgeMetrics>,
+}
+
+impl EthFinalizedDatasource {
+    pub fn new(
+        eth_sui_bridge_contract_address: String,
+        eth_rpc_url: String,
+        bridge_metrics: Arc<BridgeMetrics>,
+    ) -> Result<Self, anyhow::Error> {
+        let bridge_address = EthAddress::from_str(&eth_sui_bridge_contract_address)?;
+        Ok(Self {
+            bridge_address,
+            eth_rpc_url,
+            bridge_metrics,
+        })
+    }
+}
+
+#[async_trait]
+impl Datasource<EthData, PgBridgePersistent, ProcessedTxnData> for EthFinalizedDatasource {
+    async fn start_ingestion_task<F, M>(
+        &self,
+        task_name: String,
+        target_checkpoint: u64,
+        mut storage: PgBridgePersistent,
+        _filter: F,
+        data_mapper: M,
+    ) -> Result<(), Error>
+    where
+        F: DataFilter<EthData> + 'static,
+        M: DataMapper<EthData, ProcessedTxnData> + 'static,
+    {
+        let eth_client = Arc::new(
+            EthClient::<Http>::new(
+                &self.eth_rpc_url,
+                HashSet::from_iter(vec![self.bridge_address]),
+                self.bridge_metrics.clone(),
+            )
+            .await?,
+        );
+
+        let provider = Arc::new(
+            Provider::<Http>::try_from(&self.eth_rpc_url)?
+                .interval(std::time::Duration::from_millis(2000)),
+        );
+
+        let newest_finalized_block = storage.load(task_name.clone()).await?;
+        info!("Starting from finalized block: {}", newest_finalized_block);
+
+        let finalized_contract_addresses =
+            HashMap::from_iter(vec![(self.bridge_address, newest_finalized_block)]);
+
+        let (_task_handles, mut eth_events_rx, _) =
+            EthSyncer::new(eth_client, finalized_contract_addresses)
+                .run(self.bridge_metrics.clone())
+                .await
+                .map_err(|e| anyhow!(format!("{e:?}")))?;
+
+        'outer: while let Some((_, _, logs)) = eth_events_rx.recv().await {
+            // TODO: This for-loop can be optimzied to group tx / block info
+            // and reduce the queries issued to eth full node
+            for log in logs.iter() {
+                let block_number = log.block_number();
+                let block = provider.get_block(block_number).await?.unwrap();
+                let tx_hash = log.tx_hash();
+                let transaction = provider.get_transaction(tx_hash).await?.unwrap();
+                storage.write(data_mapper.map((log.clone(), block, transaction))?)?;
```

Comment:
1. likewise a batch_writer api will be more efficiently here
2. for max throughput, we should leverage channels to process messages in a pipeline, as opposed to blocking here